File: ios/liquid_ai/Sources/liquid_ai/LiquidAiPlugin.swift
import Flutter
import LeapModelDownloader
import LeapSDK
import UIKit

/// Flutter plugin for Liquid AI with LEAP SDK integration.
public class LiquidAiPlugin: NSObject, FlutterPlugin {
    let progressHandler = DownloadProgressHandler()
    let generationProgressHandler = GenerationProgressHandler()
    lazy var modelManager = ModelRunnerManager(progressHandler: progressHandler)
    lazy var conversationManager = ConversationManager(
        progressHandler: generationProgressHandler,
        runnerManager: modelManager
    )

    /// Method handler type alias.
    typealias MethodHandler = (FlutterMethodCall, @escaping FlutterResult) -> Void

    /// Lazy initialization of method handlers dictionary.
    private lazy var methodHandlers: [String: MethodHandler] = [
        "getPlatformVersion": { _, result in result("iOS " + UIDevice.current.systemVersion) },
        "downloadModel": handleDownloadModel,
        "loadModel": handleLoadModel,
        "loadModelFromPath": handleLoadModelFromPath,
        "unloadModel": handleUnloadModel,
        "getLoadedModelInfo": handleGetLoadedModelInfo,
        "forceUnloadAll": handleForceUnloadAll,
        "isModelDownloaded": handleIsModelDownloaded,
        "deleteModel": handleDeleteModel,
        "getCachedModels": handleGetCachedModels,
        "isModelCached": handleIsModelCached,
        "deleteAllModels": handleDeleteAllModels,
        "downloadModelFromUrl": handleDownloadModelFromUrl,
        "cancelOperation": handleCancelOperation,
        "getModelStatus": handleGetModelStatus,
        "createConversation": handleCreateConversation,
        "createConversationFromHistory": handleCreateConversationFromHistory,
        "getConversationHistory": handleGetConversationHistory,
        "disposeConversation": handleDisposeConversation,
        "clearConversationHistory": handleClearConversationHistory,
        "forkConversation": handleForkConversation,
        "exportConversation": handleExportConversation,
        "generateResponse": handleGenerateResponse,
        "stopGeneration": handleStopGeneration,
        "registerFunction": handleRegisterFunction,
        "provideFunctionResult": handleProvideFunctionResult,
        "getTokenCount": handleGetTokenCount,
    ]

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "liquid_ai",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "liquid_ai/download_progress",
            binaryMessenger: registrar.messenger()
        )
        let generationEventChannel = FlutterEventChannel(
            name: "liquid_ai/generation",
            binaryMessenger: registrar.messenger()
        )

        let instance = LiquidAiPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance.progressHandler)
        generationEventChannel.setStreamHandler(instance.generationProgressHandler)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let handler = methodHandlers[call.method] {
            handler(call, result)
        } else {
            result(FlutterMethodNotImplemented)
        }
    }
}

// MARK: - Model Handlers

extension LiquidAiPlugin {
    func handleDownloadModel(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let model = args["model"] as? String,
              let quantization = args["quantization"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: model, quantization",
                details: nil
            ))
            return
        }

        Task {
            let operationId = await modelManager.downloadModel(
                model: model,
                quantization: quantization
            )
            DispatchQueue.main.async {
                result(operationId)
            }
        }
    }

    func handleLoadModel(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let model = args["model"] as? String,
              let quantization = args["quantization"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: model, quantization",
                details: nil
            ))
            return
        }

        Task {
            let operationId = await modelManager.loadModel(
                model: model,
                quantization: quantization
            )
            DispatchQueue.main.async {
                result(operationId)
            }
        }
    }

    func handleLoadModelFromPath(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["path"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: path",
                details: nil
            ))
            return
        }

        Task {
            let operationId = await modelManager.loadModelFromPath(path: path)
            DispatchQueue.main.async {
                result(operationId)
            }
        }
    }

    func handleUnloadModel(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let runnerId = args["runnerId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: runnerId",
                details: nil
            ))
            return
        }

        Task {
            let success = await modelManager.unloadModel(runnerId: runnerId)
            DispatchQueue.main.async {
                result(success)
            }
        }
    }

    func handleGetLoadedModelInfo(_: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            let info = await modelManager.getLoadedModelInfo()
            DispatchQueue.main.async {
                result(info)
            }
        }
    }

    func handleForceUnloadAll(_: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            await modelManager.forceUnloadAll()
            DispatchQueue.main.async {
                result(nil)
            }
        }
    }

    func handleIsModelDownloaded(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let model = args["model"] as? String,
              let quantization = args["quantization"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: model, quantization",
                details: nil
            ))
            return
        }

        Task {
            let isDownloaded = await modelManager.isModelDownloaded(
                model: model,
                quantization: quantization
            )
            DispatchQueue.main.async {
                result(isDownloaded)
            }
        }
    }

    func handleDeleteModel(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let model = args["model"] as? String,
              let quantization = args["quantization"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: model, quantization",
                details: nil
            ))
            return
        }

        Task {
            do {
                try await modelManager.deleteModel(model: model, quantization: quantization)
                DispatchQueue.main.async {
                    result(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "DELETE_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }

    func handleCancelOperation(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let operationId = args["operationId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: operationId",
                details: nil
            ))
            return
        }

        Task {
            await modelManager.cancelOperation(operationId: operationId)
            DispatchQueue.main.async {
                result(nil)
            }
        }
    }

    func handleGetModelStatus(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let model = args["model"] as? String,
              let quantization = args["quantization"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: model, quantization",
                details: nil
            ))
            return
        }

        Task {
            let status = await modelManager.getModelStatus(
                model: model,
                quantization: quantization
            )
            let statusMap: [String: Any] = [
                "type": status.type.description,
                "progress": status.progress,
            ]
            DispatchQueue.main.async {
                result(statusMap)
            }
        }
    }

    func handleGetCachedModels(_: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            let models = await modelManager.getCachedModels()
            DispatchQueue.main.async {
                result(models)
            }
        }
    }

    func handleIsModelCached(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let modelId = args["modelId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: modelId",
                details: nil
            ))
            return
        }

        Task {
            let isCached = await modelManager.isModelCached(modelId: modelId)
            DispatchQueue.main.async {
                result(isCached)
            }
        }
    }

    func handleDeleteAllModels(_: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            do {
                try await modelManager.deleteAllModels()
                DispatchQueue.main.async {
                    result(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "DELETE_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }

    func handleDownloadModelFromUrl(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let url = args["url"] as? String,
              let modelId = args["modelId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: url, modelId",
                details: nil
            ))
            return
        }

        let quantization = args["quantization"] as? String ?? "custom"

        Task {
            let operationId = await modelManager.downloadModelFromUrl(
                url: url,
                modelId: modelId,
                quantization: quantization
            )
            DispatchQueue.main.async {
                result(operationId)
            }
        }
    }

}

// MARK: - Conversation Handlers

extension LiquidAiPlugin {
    func handleCreateConversation(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let runnerId = args["runnerId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: runnerId",
                details: nil
            ))
            return
        }

        let systemPrompt = args["systemPrompt"] as? String

        Task {
            do {
                let conversationId = try await conversationManager.createConversation(
                    runnerId: runnerId,
                    systemPrompt: systemPrompt
                )
                DispatchQueue.main.async {
                    result(conversationId)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "CREATE_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }

    func handleCreateConversationFromHistory(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let runnerId = args["runnerId"] as? String,
              let history = args["history"] as? [[String: Any]] else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: runnerId, history",
                details: nil
            ))
            return
        }

        Task {
            do {
                let conversationId = try await conversationManager.createConversationFromHistory(
                    runnerId: runnerId,
                    history: history
                )
                DispatchQueue.main.async {
                    result(conversationId)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "CREATE_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }

    func handleGetConversationHistory(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let conversationId = args["conversationId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: conversationId",
                details: nil
            ))
            return
        }

        Task {
            do {
                let history = try await conversationManager.getHistory(conversationId: conversationId)
                DispatchQueue.main.async {
                    result(history)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "GET_HISTORY_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }

    func handleDisposeConversation(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let conversationId = args["conversationId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: conversationId",
                details: nil
            ))
            return
        }

        Task {
            await conversationManager.disposeConversation(conversationId)
            DispatchQueue.main.async {
                result(nil)
            }
        }
    }

    func handleClearConversationHistory(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let conversationId = args["conversationId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: conversationId",
                details: nil
            ))
            return
        }

        let keepSystemPrompt = args["keepSystemPrompt"] as? Bool ?? true

        Task {
            do {
                try await conversationManager.clearHistory(
                    conversationId: conversationId,
                    keepSystemPrompt: keepSystemPrompt
                )
                DispatchQueue.main.async {
                    result(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "CLEAR_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }

    func handleForkConversation(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let conversationId = args["conversationId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: conversationId",
                details: nil
            ))
            return
        }

        Task {
            do {
                let newConversationId = try await conversationManager.forkConversation(conversationId)
                DispatchQueue.main.async {
                    result(newConversationId)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "FORK_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }

    func handleExportConversation(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let conversationId = args["conversationId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: conversationId",
                details: nil
            ))
            return
        }

        Task {
            do {
                let json = try await conversationManager.exportConversation(conversationId)
                DispatchQueue.main.async {
                    result(json)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "EXPORT_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }
}

// MARK: - Generation Handlers

extension LiquidAiPlugin {
    func handleGenerateResponse(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let conversationId = args["conversationId"] as? String,
              let message = args["message"] as? [String: Any] else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: conversationId, message",
                details: nil
            ))
            return
        }

        let options = args["options"] as? [String: Any]

        Task {
            do {
                let generationId = try await conversationManager.generateResponse(
                    conversationId: conversationId,
                    message: message,
                    options: options
                )
                DispatchQueue.main.async {
                    result(generationId)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "GENERATION_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }

    func handleStopGeneration(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let generationId = args["generationId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: generationId",
                details: nil
            ))
            return
        }

        Task {
            await conversationManager.stopGeneration(generationId)
            DispatchQueue.main.async {
                result(nil)
            }
        }
    }
}

// MARK: - Function Calling Handlers

extension LiquidAiPlugin {
    func handleRegisterFunction(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let conversationId = args["conversationId"] as? String,
              let function = args["function"] as? [String: Any] else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: conversationId, function",
                details: nil
            ))
            return
        }

        Task {
            do {
                try await conversationManager.registerFunction(
                    conversationId: conversationId,
                    function: function
                )
                DispatchQueue.main.async {
                    result(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "REGISTER_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }

    func handleProvideFunctionResult(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let conversationId = args["conversationId"] as? String,
              let functionResult = args["result"] as? [String: Any] else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: conversationId, result",
                details: nil
            ))
            return
        }

        Task {
            do {
                try await conversationManager.provideFunctionResult(
                    conversationId: conversationId,
                    result: functionResult
                )
                DispatchQueue.main.async {
                    result(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "FUNCTION_RESULT_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }
}

// MARK: - Token Counting Handlers

extension LiquidAiPlugin {
    func handleGetTokenCount(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let conversationId = args["conversationId"] as? String else
        {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required argument: conversationId",
                details: nil
            ))
            return
        }

        Task {
            do {
                let tokenCount = try await conversationManager.getTokenCount(conversationId: conversationId)
                DispatchQueue.main.async {
                    result(tokenCount)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "TOKEN_COUNT_FAILED",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }
}

// MARK: - ModelDownloadStatusType Extension

extension ModelDownloader.ModelDownloadStatusType {
    var description: String {
        switch self {
        case .notOnLocal:
            "notOnLocal"
        case .downloadInProgress:
            "downloadInProgress"
        case .downloaded:
            "downloaded"
        }
    }
}

File: android/src/main/kotlin/dev/sogl/liquid_ai/LiquidAiPlugin.kt
package dev.sogl.liquid_ai

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*

/// Flutter plugin for Liquid AI with LEAP SDK integration.
class LiquidAiPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var generationEventChannel: EventChannel
    private val progressHandler = DownloadProgressHandler()
    private val generationProgressHandler = GenerationProgressHandler()
    private lateinit var modelManager: ModelRunnerManager
    private lateinit var conversationManager: ConversationManager
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "liquid_ai")
        channel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "liquid_ai/download_progress")
        eventChannel.setStreamHandler(progressHandler)

        generationEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "liquid_ai/generation")
        generationEventChannel.setStreamHandler(generationProgressHandler)

        modelManager = ModelRunnerManager(progressHandler)
        conversationManager = ConversationManager(generationProgressHandler, modelManager)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "downloadModel" -> handleDownloadModel(call, result)
            "loadModel" -> handleLoadModel(call, result)
            "unloadModel" -> handleUnloadModel(call, result)
            "isModelDownloaded" -> handleIsModelDownloaded(call, result)
            "deleteModel" -> handleDeleteModel(call, result)
            "cancelOperation" -> handleCancelOperation(call, result)
            "getModelStatus" -> handleGetModelStatus(call, result)
            // Conversation Management
            "createConversation" -> handleCreateConversation(call, result)
            "createConversationFromHistory" -> handleCreateConversationFromHistory(call, result)
            "getConversationHistory" -> handleGetConversationHistory(call, result)
            "disposeConversation" -> handleDisposeConversation(call, result)
            "exportConversation" -> handleExportConversation(call, result)
            // Generation
            "generateResponse" -> handleGenerateResponse(call, result)
            "stopGeneration" -> handleStopGeneration(call, result)
            // Function Calling
            "registerFunction" -> handleRegisterFunction(call, result)
            "provideFunctionResult" -> handleProvideFunctionResult(call, result)
            // Token Counting
            "getTokenCount" -> handleGetTokenCount(call, result)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        generationEventChannel.setStreamHandler(null)
        conversationManager.dispose()
        modelManager.dispose()
        scope.cancel()
    }

    // MARK: - Method Handlers

    private fun handleDownloadModel(call: MethodCall, result: Result) {
        val model = call.argument<String>("model")
        val quantization = call.argument<String>("quantization")

        if (model == null || quantization == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: model, quantization",
                null
            )
            return
        }

        scope.launch {
            val operationId = withContext(Dispatchers.IO) {
                modelManager.downloadModel(model, quantization)
            }
            result.success(operationId)
        }
    }

    private fun handleLoadModel(call: MethodCall, result: Result) {
        val model = call.argument<String>("model")
        val quantization = call.argument<String>("quantization")

        if (model == null || quantization == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: model, quantization",
                null
            )
            return
        }

        scope.launch {
            val operationId = withContext(Dispatchers.IO) {
                modelManager.loadModel(model, quantization)
            }
            result.success(operationId)
        }
    }

    private fun handleUnloadModel(call: MethodCall, result: Result) {
        val runnerId = call.argument<String>("runnerId")

        if (runnerId == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required argument: runnerId",
                null
            )
            return
        }

        scope.launch {
            val success = withContext(Dispatchers.IO) {
                modelManager.unloadModel(runnerId)
            }
            result.success(success)
        }
    }

    private fun handleIsModelDownloaded(call: MethodCall, result: Result) {
        val model = call.argument<String>("model")
        val quantization = call.argument<String>("quantization")

        if (model == null || quantization == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: model, quantization",
                null
            )
            return
        }

        scope.launch {
            val isDownloaded = withContext(Dispatchers.IO) {
                modelManager.isModelDownloaded(model, quantization)
            }
            result.success(isDownloaded)
        }
    }

    private fun handleDeleteModel(call: MethodCall, result: Result) {
        val model = call.argument<String>("model")
        val quantization = call.argument<String>("quantization")

        if (model == null || quantization == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: model, quantization",
                null
            )
            return
        }

        scope.launch {
            try {
                withContext(Dispatchers.IO) {
                    modelManager.deleteModel(model, quantization)
                }
                result.success(null)
            } catch (e: Exception) {
                result.error(
                    "DELETE_FAILED",
                    e.message ?: "Delete failed",
                    null
                )
            }
        }
    }

    private fun handleCancelOperation(call: MethodCall, result: Result) {
        val operationId = call.argument<String>("operationId")

        if (operationId == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required argument: operationId",
                null
            )
            return
        }

        modelManager.cancelOperation(operationId)
        result.success(null)
    }

    private fun handleGetModelStatus(call: MethodCall, result: Result) {
        val model = call.argument<String>("model")
        val quantization = call.argument<String>("quantization")

        if (model == null || quantization == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: model, quantization",
                null
            )
            return
        }

        scope.launch {
            val status = withContext(Dispatchers.IO) {
                modelManager.getModelStatus(model, quantization)
            }
            result.success(status)
        }
    }

    // MARK: - Conversation Handlers

    private fun handleCreateConversation(call: MethodCall, result: Result) {
        val runnerId = call.argument<String>("runnerId")

        if (runnerId == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required argument: runnerId",
                null
            )
            return
        }

        val systemPrompt = call.argument<String>("systemPrompt")

        scope.launch {
            try {
                val conversationId = withContext(Dispatchers.IO) {
                    conversationManager.createConversation(runnerId, systemPrompt)
                }
                result.success(conversationId)
            } catch (e: Exception) {
                result.error(
                    "CREATE_FAILED",
                    e.message ?: "Create conversation failed",
                    null
                )
            }
        }
    }

    private fun handleCreateConversationFromHistory(call: MethodCall, result: Result) {
        val runnerId = call.argument<String>("runnerId")
        @Suppress("UNCHECKED_CAST")
        val history = call.argument<List<Map<String, Any>>>("history")

        if (runnerId == null || history == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: runnerId, history",
                null
            )
            return
        }

        scope.launch {
            try {
                val conversationId = withContext(Dispatchers.IO) {
                    conversationManager.createConversationFromHistory(runnerId, history)
                }
                result.success(conversationId)
            } catch (e: Exception) {
                result.error(
                    "CREATE_FAILED",
                    e.message ?: "Create conversation failed",
                    null
                )
            }
        }
    }

    private fun handleGetConversationHistory(call: MethodCall, result: Result) {
        val conversationId = call.argument<String>("conversationId")

        if (conversationId == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required argument: conversationId",
                null
            )
            return
        }

        scope.launch {
            try {
                val history = withContext(Dispatchers.IO) {
                    conversationManager.getHistory(conversationId)
                }
                result.success(history)
            } catch (e: Exception) {
                result.error(
                    "GET_HISTORY_FAILED",
                    e.message ?: "Get history failed",
                    null
                )
            }
        }
    }

    private fun handleDisposeConversation(call: MethodCall, result: Result) {
        val conversationId = call.argument<String>("conversationId")

        if (conversationId == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required argument: conversationId",
                null
            )
            return
        }

        conversationManager.disposeConversation(conversationId)
        result.success(null)
    }

    private fun handleExportConversation(call: MethodCall, result: Result) {
        val conversationId = call.argument<String>("conversationId")

        if (conversationId == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required argument: conversationId",
                null
            )
            return
        }

        scope.launch {
            try {
                val json = withContext(Dispatchers.IO) {
                    conversationManager.exportConversation(conversationId)
                }
                result.success(json)
            } catch (e: Exception) {
                result.error(
                    "EXPORT_FAILED",
                    e.message ?: "Export failed",
                    null
                )
            }
        }
    }

    // MARK: - Generation Handlers

    private fun handleGenerateResponse(call: MethodCall, result: Result) {
        val conversationId = call.argument<String>("conversationId")
        @Suppress("UNCHECKED_CAST")
        val message = call.argument<Map<String, Any>>("message")

        if (conversationId == null || message == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: conversationId, message",
                null
            )
            return
        }

        @Suppress("UNCHECKED_CAST")
        val options = call.argument<Map<String, Any>>("options")

        scope.launch {
            try {
                val generationId = withContext(Dispatchers.IO) {
                    conversationManager.generateResponse(conversationId, message, options)
                }
                result.success(generationId)
            } catch (e: Exception) {
                result.error(
                    "GENERATION_FAILED",
                    e.message ?: "Generation failed",
                    null
                )
            }
        }
    }

    private fun handleStopGeneration(call: MethodCall, result: Result) {
        val generationId = call.argument<String>("generationId")

        if (generationId == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required argument: generationId",
                null
            )
            return
        }

        conversationManager.stopGeneration(generationId)
        result.success(null)
    }

    // MARK: - Function Calling Handlers

    private fun handleRegisterFunction(call: MethodCall, result: Result) {
        val conversationId = call.argument<String>("conversationId")
        @Suppress("UNCHECKED_CAST")
        val function = call.argument<Map<String, Any>>("function")

        if (conversationId == null || function == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: conversationId, function",
                null
            )
            return
        }

        try {
            conversationManager.registerFunction(conversationId, function)
            result.success(null)
        } catch (e: Exception) {
            result.error(
                "REGISTER_FAILED",
                e.message ?: "Register function failed",
                null
            )
        }
    }

    private fun handleProvideFunctionResult(call: MethodCall, result: Result) {
        val conversationId = call.argument<String>("conversationId")
        @Suppress("UNCHECKED_CAST")
        val functionResult = call.argument<Map<String, Any>>("result")

        if (conversationId == null || functionResult == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: conversationId, result",
                null
            )
            return
        }

        try {
            conversationManager.provideFunctionResult(conversationId, functionResult)
            result.success(null)
        } catch (e: Exception) {
            result.error(
                "FUNCTION_RESULT_FAILED",
                e.message ?: "Provide function result failed",
                null
            )
        }
    }

    // MARK: - Token Counting

    private fun handleGetTokenCount(call: MethodCall, result: Result) {
        // Token counting API is not available on Android SDK
        result.error(
            "UNSUPPORTED",
            "Token counting is not supported on Android. This feature is only available on iOS.",
            null
        )
    }
}
